### PR TITLE
Refactor client generation to take gateway as arg.

### PR DESCRIPTION
**Pull request by Raynos**
This prep allows us to get access to the `gateway.Logger`
in follow up PRs where we add a `HTTPClientRequest` class.

r: @uber/zanzibar-team

### CODE REVIEW

File: codegen/templates/http_client.tmpl [hunk 1]
```diff
@@ -67,15 +64,15 @@ func (c *{{$clientName}}) {{title .Name}}(ctx context.Context, r *{{.RequestType
 		return nil, err
 	}
 	req.Header.Set("Content-Type", "application/json")
-	return c.Client.Do(req.WithContext(ctx))
+	return c.client.Client.Do(req.WithContext(ctx))
```
Comment (zhenghuiwang) [inline codegen/templates/http_client.tmpl]: Nit:`c.client.Client` repeats `client`. using 'zanzibar.HTTPClient' as anonymous field could remove one word 'client'.

Comment (Raynos) [inline codegen/templates/http_client.tmpl]: Good point, I'd rather avoid inheriting http.Client as public API for now. 

naming these things is a bit hard.

**Comment by ChuntaoLu** [comment]
Is passing in a gateway pointer necessary if all we need is logger access? Do we predict other use cases with gateway other than logger?

**Comment by Raynos** [comment]
@ChuntaoLu `gateway.MetricsScope` as well, its used for logger, metrics, jaeger tracing, etc.

**Pull request merged**